executor/benchmark_test.go

```diff
@@ -0,0 +1,344 @@
+}
+
+func (a aggTestCase) String() string {
+	return fmt.Sprintf("(%v|%v|%v|%v|%v|%v)",
+		a.execType, a.aggFunc, a.hasDistinct, a.rows, a.groupByNDV, a.concurrency)
+}
+
+func defaultAggTestCase(exec string) *aggTestCase {
+	ctx := mock.NewContext()
+	ctx.GetSessionVars().InitChunkSize = 1024
+	ctx.GetSessionVars().MaxChunkSize = 1024
+	return &aggTestCase{exec, ast.AggFuncSum, 1000, false, 10000000, 4, ctx}
+}
+
+func buildHashAggExecutor(ctx sessionctx.Context, src Executor, schema *expression.Schema,
+	aggFuncs []*aggregation.AggFuncDesc, groupItems []expression.Expression) Executor {
+	plan := new(core.PhysicalHashAgg)
+	plan.AggFuncs = aggFuncs
+	plan.GroupByItems = groupItems
+	plan.SetSchema(schema)
+	plan.Init(ctx, nil)
+	plan.SetChildren(nil)
+	b := newExecutorBuilder(ctx, nil)
+	exec := b.build(plan)
+	hashAgg := exec.(*HashAggExec)
+	hashAgg.children[0] = src
+	return exec
+}
+
+func buildStreamAggExecutor(ctx sessionctx.Context, src Executor, schema *expression.Schema,
+	aggFuncs []*aggregation.AggFuncDesc, groupItems []expression.Expression) Executor {
+	plan := new(core.PhysicalStreamAgg)
+	plan.AggFuncs = aggFuncs
+	plan.GroupByItems = groupItems
+	plan.SetSchema(schema)
+	plan.Init(ctx, nil)
+	plan.SetChildren(nil)
+	b := newExecutorBuilder(ctx, nil)
+	exec := b.build(plan)
+	streamAgg := exec.(*StreamAggExec)
+	streamAgg.children[0] = src
+	return exec
+}
+
+func buildAggExecutor(b *testing.B, cas *aggTestCase, child Executor) Executor {
+	ctx := cas.ctx
+	if err := ctx.GetSessionVars().SetSystemVar(variable.TiDBHashAggFinalConcurrency, fmt.Sprintf("%v", cas.concurrency)); err != nil {
+		b.Fatal(err)
+	}
+	if err := ctx.GetSessionVars().SetSystemVar(variable.TiDBHashAggPartialConcurrency, fmt.Sprintf("%v", cas.concurrency)); err != nil {
+		b.Fatal(err)
+	}
+
+	childCols := cas.columns()
+	schema := expression.NewSchema(childCols...)
+	groupBy := []expression.Expression{childCols[1]}
+	aggFunc := aggregation.NewAggFuncDesc(cas.ctx, cas.aggFunc, []expression.Expression{childCols[0]}, cas.hasDistinct)
+	aggFuncs := []*aggregation.AggFuncDesc{aggFunc}
+
+	var aggExec Executor
+	switch cas.execType {
+	case "hash":
+		aggExec = buildHashAggExecutor(cas.ctx, child, schema, aggFuncs, groupBy)
+	case "stream":
+		aggExec = buildStreamAggExecutor(cas.ctx, child, schema, aggFuncs, groupBy)
+	default:
+		b.Fatal("not implement")
+	}
+	return aggExec
+}
+
+func benchmarkAggExecWithCase(b *testing.B, cas *aggTestCase) {
+	cols := cas.columns()
+	orders := []bool{false, cas.execType == "stream"}
+	dataSource := buildMockDataSource(mockDataSourceParameters{
+		schema: expression.NewSchema(cols...),
+		ndvs:   []int{0, cas.groupByNDV},
+		orders: orders,
+		rows:   cas.rows,
+		ctx:    cas.ctx,
+	})
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		b.StopTimer() // prepare a new agg-executor
+		aggExec := buildAggExecutor(b, cas, dataSource)
+		tmpCtx := context.Background()
+		chk := aggExec.newFirstChunk()
+		dataSource.prepareChunks()
```

Comment:
I think we should move this out of the benchmark loop. It consumes a lot of CPU and Memory time.

Reply:
OK